### DISCUSSION TITLE
Getting new FCM Token on every new app launch ( iOS - After SWIPE AWAY the app)

### DISCUSSION THREAD

**Discussion by TommyLeong** [invertase/react-native-firebase#4441] [Q&A] [answered]
Hello all, 

Understand that, by default there's a auto registration for iOS for remote message. Hence, for us to get the FCM token and pass back to our backend server, we'll just call `await messaging().getToken()` ?

If that's how it work, im guess im doing right so far. But I'm getting a new FCM Token on every app launch which is kinda weird right? 

Suppose I'm expecting a FCM token should only get renew after deleting and reinstalling the app. Please help to correct me if I'm wrong. 🙏 

Thank you

**Comment by TommyLeong** [comment]
Then I realize the method [didReceiveRegistrationToken](https://github.com/invertase/react-native-firebase/blob/d3796effe87040bad61355202104cdbbf169545e/packages/messaging/ios/RNFBMessaging/RNFBMessaging%2BFIRMessagingDelegate.m#L50) was called on app launch. I put a breakpoint here to debug..

So the method was entered twice. 
- On the first round, when I try to print out the `fcmToken`, it returns me the "existing FCM token".
- On the second time, when I try to print out the `fcmToken`, it returns me a new FCM token

😞

**Reply by mikehardy** [reply]
This seems like an issue not a discussion, and I feel like it exists somewhere already but can't find it.
Either way, I have no idea what versions you are running, so hard to triage.

Related: you must integrate this if you are working with tokens right now, and reproduce again as I think it could have impact: https://github.com/invertase/react-native-firebase/issues/4299#issuecomment-705128266

General thought: what's the stack on the second entry? Who called it and why? It's possible from what I know based on what you have provided that you are calling invalidate somewhere. Not saying you are, but as a troubleshooter, I don't know

**Reply by mikehardy** [reply]
Here's the related entry #4390 / cross linking @ayon-affle

**Reply by TommyLeong** [reply]
I'm uncertain whether this is an issue from RNFirebase, so I decided to discuss here before creating a ticket. You're right, I'm suppose to provide the versioning i'm using.
```
    "@react-native-firebase/analytics": "^7.3.1",
    "@react-native-firebase/app": "^8.3.0",
    "@react-native-firebase/crashlytics": "^8.1.2",
    "@react-native-firebase/messaging": "^7.8.4",
```

I look through my code, I can't find any method that's performing any invalidation anywhere. When you say invalidate, i assume you are saying `messaging().deleteToken()`.

Back to what you mentioned about https://github.com/invertase/react-native-firebase/issues/4299#issuecomment-705128266
You mean to say I should include the following in my podfile? 🤔🤔🤔 
```
 # Set up Firebase SDK version
  $FirebaseSDKVersion = '6.34.0'
  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => $FirebaseSDKVersion
  pod 'FirebaseInstanceID', :git => 'https://github.com/invertase/firebase-ios-sdk.git', :branch => 'release-6.34.0-patched'
```

**Reply by mikehardy** [reply]
the FirebaseFirestore reference is optional - it's an optimization of build times for Firestore users. the FirebaseInstanceID line though, definitely - that's the trick

**Reply by mikehardy** [reply]
and your versions are out of date

**Reply by TommyLeong** [reply]
Adding both them in don't solve the issue for me. Perhaps this is a different issue to what you are linking mine too. Yes, i did follow this `cd ios && rm -f Podfile.lock && pod install --repo-update`.
As soon as I kill the app, and relaunching it. I'm getting a new FCM Token again.

**Reply by TommyLeong** [reply]
I'm still uncertain on the root cause on what causing it to refresh my app a new FCM Token. However from what I troubleshoot, `com.firebase.iid.notif.refresh-token` was the key that trigger this refresh. It seems to trigger the observer that was appended earlier added by method `setupNotificationListeners`

![image](https://user-images.githubusercontent.com/9990179/97032995-3943ff00-1595-11eb-8d97-47b64dd5553f.png)

**Comment by TommyLeong** [comment, marked as answer]
**Solved.**
To wrap up this issue...

Earlier, the issue for this "new FCM Token" on every app launch (after SWIPE AWAY the app). It seems to be a conflict between RNFirebase SDK with another 3rd party SDK of mine in project. 

The 3rd party SDK did certain action on 
- Firebase App Configuration
- passing APNS token to Firebase
- etcs (which I'm not aware of, since it's a framework..I'm unable to read the code inside)

They then removed the "Firebase part" within their SDK, turns out.. the FCM token behave as it is. No longer changing on every app launch. 

Thanks Mike for joining the discussion! Appreciate 👍

**Reply by mikehardy** [reply]
Glad to hear this is resolved, and now matches my work project - I have stable tokens
